feat(node/child_process): `spawnSync`

iuioiua opened the pull request:
Closes #1654. Passing tests are:
- [x] `test-child-process-spawnsync-args.js`**
- [x] `test-child-process-spawnsync-env.js`
- [ ] `test-child-process-spawnsync-input.js`
- [ ] `test-child-process-spawnsync-kill-signal.js`*
- [x] `test-child-process-spawnsync-maxbuf.js`
- [ ] `test-child-process-spawnsync-shell.js`
- [ ] `test-child-process-spawnsync-timeout.js`*
- [ ] `test-child-process-spawnsync-validation-errors.js`
- [x] `test-child-process-spawnsync.js`

*`Deno.spawnSync()` actually doesn't support `signal` as input. So I think this test cannot be catered for.
** This is currently failing with a `'\"cd\"' is not recognized as an internal or external command` error, as seen from `stderr`, due to an escaping issue. So, this test has been disabled for Windows until https://github.com/denoland/deno/issues/8852 is resolved.
https://github.com/denoland/deno_std/blob/6a521ac65e049c318bc205cb5a76de7ac818c546/node/internal/child_process.ts#L444-L450

cjihrig commented on node/child_process_test.ts:
```diff
@@ -507,3 +508,91 @@ Deno.test({
   },
 });
+
+Deno.test("[node/child_process] spawnSync()", async (t) => {
```
Instead of introducing our own tests, can we vendor more of the Node.js test suite for this functionality.

iuioiua replied:
Yeah, haha. I thought the same after realising the amount of testing I'd be doing is ridiculous.

iuioiua commented:
@cjihrig, I'm thinking that getting `spawnSync` to a point where `test-child-process-spawnsync-input.js` and `test-child-process-spawnsync-validation-errors.js` are passing would suffice to push this PR. The other tests can be worked on later. What do you think?

iuioiua commented:
> @cjihrig, I'm thinking that getting `spawnSync` to a point where `test-child-process-spawnsync-input.js` and `test-child-process-spawnsync-validation-errors.js` are passing would suffice to push this PR. The other tests can be worked on later. What do you think?

Actually, we use TS. `test-child-process-spawnsync-validation-errors.js` shouldn't have to be tested, right?

cjihrig commented:
> I'm thinking that getting spawnSync to a point where test-child-process-spawnsync-input.js and test-child-process-spawnsync-validation-errors.js are passing would suffice to push this PR.

I'm not opposed to that idea, but how/why did you settle on those two tests? Why not `test/parallel/test-child-process-spawnsync.js`, for example?

> Actually, we use TS. test-child-process-spawnsync-validation-errors.js shouldn't have to be tested, right?

I think we should port that test. The errors that it validates are Node specific errors, and it's still possible to trigger that code.

iuioiua commented:
> > I'm thinking that getting spawnSync to a point where test-child-process-spawnsync-input.js and test-child-process-spawnsync-validation-errors.js are passing would suffice to push this PR.
> 
> I'm not opposed to that idea, but how/why did you settle on those two tests? Why not `test/parallel/test-child-process-spawnsync.js`, for example?

My question is: how much functionality will suffice in having `spawnSync` ready for use (i.e. able to work in most cases)? I thought those might be the two remaining failing tests that should be fixed. `test-child-process-spawnsync.js` is already passing.

In my initial comment for this PR, I've listed the passing and remaining tests. One could argue that `spawnSync` is ready for use now. What do you think?

> 
> > Actually, we use TS. test-child-process-spawnsync-validation-errors.js shouldn't have to be tested, right?
> 
> I think we should port that test. The errors that it validates are Node specific errors, and it's still possible to trigger that code.

Some assertions within that test simply type-check. However, those assertions are failing even though types are defined. Must I include type-checking within my TypeScript code?

cjihrig commented:
Sorry, I missed your list of already passing tests. I agree that this is probably enough functionality to merge and iterate on. Can you rebase this and regenerate the tests (we're now up to Node 18).

I won't hold up this PR over it, but I strongly believe we should get the type checking code in place.

iuioiua commented:
> Sorry, I missed your list of already passing tests. I agree that this is probably enough functionality to merge and iterate on. Can you rebase this and regenerate the tests (we're now up to Node 18).
> 
> I won't hold up this PR over it, but I strongly believe we should get the type checking code in place.

I've submitted the PR as I'd prefer to land the function sooner. Thanks for your help!

cjihrig commented on node/_tools/test/parallel/test-child-process-spawnsync-maxbuf.js:
```diff
@@ -0,0 +1,65 @@
+
+// Verify that an error is returned if maxBuffer is surpassed.
+{
+  const ret = spawnSync(Deno.execPath(), args, { maxBuffer: 1 });
```
Is there a problem with `process.execPath`? It seems to be implemented.

iuioiua replied:
I thought `process` properties weren't working. All good, fixed now.

cjihrig commented on node/_tools/test/parallel/test-child-process-spawnsync-env.js:
```diff
@@ -0,0 +1,44 @@
+
+  const expected = 'bar';
+  const child = cp.spawnSync(Deno.execPath(), ["run", "-A", "--unstable", __filename, 'child'], {
+    env: Object.assign(Deno.env.toObject(), { foo: expected })
```
Same question about using the `Deno` object in this file. Those `process` APIs seem to be available.

iuioiua replied:
Fixed now. Thanks for pointing out.

cjihrig commented on node/internal/util.mjs:
```diff
@@ -143,9 +143,23 @@ export function promisify(
 }
 
+export function convertToValidSignal(signal) {
+  if (typeof signal === "number" && getSignalsToNamesMapping()[signal]) {
```
`getSignalsToNamesMapping` doesn't seem to be defined.

cjihrig commented on node/internal/util.mjs:
```diff
@@ -143,9 +143,23 @@ export function promisify(
+  }
+
+  if (typeof signal === "string") {
+    const signalName = signals[StringPrototypeToUpperCase(signal)];
```
`StringPrototypeToUpperCase` doesn't seem to be defined.

cjihrig commented on node/internal/util.mjs:
```diff
@@ -143,9 +143,23 @@ export function promisify(
+    if (signalName) return signalName;
+  }
+
+  throw new ERR_UNKNOWN_SIGNAL(signal);
```
`ERR_UNKNOWN_SIGNAL` also doesn't appear to be defined in this file.

cjihrig commented on node/child_process.ts:
```diff
@@ -140,6 +143,49 @@ export function spawn(
 }
 
+function validateTimeout(timeout?: number) {
+  if (timeout !== undefined && !(Number.isInteger(timeout) && timeout >= 0)) {
```
To match the Node implementation:

```suggestion
  if (timeout != null && !(Number.isInteger(timeout) && timeout >= 0)) {
```

cjihrig commented on node/child_process.ts:
```diff
@@ -140,6 +143,49 @@ export function spawn(
+
+function validateMaxBuffer(maxBuffer?: number) {
+  if (
+    maxBuffer !== undefined &&
```
To match the Node implementation:

```suggestion
    maxBuffer != null &&
```

cjihrig commented on node/child_process.ts:
```diff
@@ -140,6 +143,49 @@ export function spawn(
+  command: string,
+  argsOrOptions?: string[] | SpawnSyncOptions,
+  maybeOptions?: SpawnSyncOptions,
+): SpawnSyncResult {
```
Can you add some code to throw if a valid, but currently unsupported option is passed in.

cjihrig commented on node/internal/child_process.ts:
```diff
@@ -470,3 +474,100 @@ function buildCommand(
+    gid,
+    maxBuffer,
+  } = options;
+  const normalisezedStdio = normalizeStdioOption(stdio);
```
Typo: `normalizedStdio`

cjihrig commented on node/internal/child_process.ts:
```diff
@@ -470,3 +474,100 @@ function buildCommand(
+
+    const { signal } = output;
+    const status = signal ? null : 0;
+    let stdout = Buffer.from(output.stdout) as string | Buffer;
+    let stderr = Buffer.from(output.stderr) as string | Buffer;
+
+    if (stdout.length > maxBuffer! || stderr.length > maxBuffer!) {
+      result.error = _createSpawnSyncError("ENOBUFS", command, args);
+    }
+
+    if (encoding && encoding !== "buffer") {
+      stdout = stdout.toString(encoding);
+      stderr = stderr.toString(encoding);
+    }
```
Is it possible to avoid converting to a Buffer and then potentially converting to a string? It would be nice if we only needed to do one conversion.

iuioiua commented on node/_tools/config.json:
```diff
@@ -17,6 +17,8 @@
       "test-buffer-from.js",
       "test-buffer-includes.js",
       "test-buffer-indexof.js",
+      "test-child-process-spawnsync-env.js",
+      "test-child-process-spawnsync-maxbuf.js",
```
`test-child-process-spawnsync-maxbuf.js` uses `-e` for eval, whereas Deno uses `eval`.

Same reason for `test-child-process-spawnsync-env.js` on line 40. See other reasons below.

cjihrig commented on node/_tools/test/parallel/test-child-process-spawnsync-env.js:
```diff
@@ -0,0 +1,45 @@
+
+
+if (Deno.args[0] === 'child') {
+  console.log(Deno.env.get("foo"));
```
Please use `process.env` instead of `Deno.env.get()`.

cjihrig commented on node/_tools/test/parallel/test-child-process-spawnsync-env.js:
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+
+if (Deno.args[0] === 'child') {
```
Please use `process.argv` instead of `Deno.args`.

cjihrig commented:
@iuioiua there are still a few unaddressed comments. Would you mind taking a look. Ideally, I'd like to get this merged this week.

iuioiua commented:
> @iuioiua there are still a few unaddressed comments. Would you mind taking a look. Ideally, I'd like to get this merged this week.

Yeah, sorry. I haven't had many chances to work on this in the last week. Yeah, I'll get this done this week.

Are you able to take a look at some of the comments I've made as well?

cjihrig commented:
> Are you able to take a look at some of the comments I've made as well?

Sure. Any comments in particular? From what I can tell, all of your comments have been addressed or are now outdated due to changes in the diff.

iuioiua commented:
> > Are you able to take a look at some of the comments I've made as well?
> 
> Sure. Any comments in particular? From what I can tell, all of your comments have been addressed or are now outdated due to changes in the diff.

My bad. I just discovered the "submit review" functionality for PRs 😂

iuioiua commented:
Accidentally closed this PR. Re-opening...

cjihrig reviewed (approved):
After reviewing again, I think this is in good enough shape to merge and iterate on. Thanks for the contribution.

iuioiua commented:
> After reviewing again, I think this is in good enough shape to merge and iterate on. Thanks for the contribution.

Glad to see! I wish I had handled everything in a much cleaner/faster manner, but I’m learning. Thanks for your help!
